File: backend/src/lib/findProjects.js
import fs from 'node:fs/promises';
import path from 'node:path';

export const PROJECT_MARKERS = {
  'pom.xml': 'maven',
  'package.json': 'npm',
  'Cargo.toml': 'rust',
  'go.mod': 'go',
};

const DEFAULT_IGNORE_DIRS = [
  'node_modules',
  'target',
  'dist',
  'build',
  'out',
  '.git',
  'venv',
  '.venv',
  'env',
  '__pycache__',
];

// Cherche récursivement, à partir de rootDir, les répertoires contenant un
// fichier marqueur de projet (pom.xml, package.json, Cargo.toml, go.mod).
// Une fois un marqueur trouvé dans un répertoire, ses sous-répertoires ne
// sont pas explorés.
export async function findProjects(rootDir, { extraIgnoreDirs = [] } = {}) {
  const ignoreDirs = new Set([...DEFAULT_IGNORE_DIRS, ...extraIgnoreDirs]);
  const results = [];
  await walk(rootDir, ignoreDirs, results);
  return results;
}

async function walk(dir, ignoreDirs, results) {
  let entries;
  try {
    entries = await fs.readdir(dir, { withFileTypes: true });
  } catch {
    return;
  }

  const foundMarkers = entries
    .filter((entry) => entry.isFile() && entry.name in PROJECT_MARKERS)
    .map((entry) => entry.name);

  if (foundMarkers.length > 0) {
    results.push({ dir, files: foundMarkers });
    return;
  }

  const subDirs = entries.filter((entry) => entry.isDirectory() && !ignoreDirs.has(entry.name));
  for (const subDir of subDirs) {
    await walk(path.join(dir, subDir.name), ignoreDirs, results);
  }
}

File: backend-java/src/main/java/com/simpledash/lib/FindProjects.java
package com.simpledash.lib;

import java.io.IOException;
import java.nio.file.DirectoryStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

// Portage de backend/src/lib/findProjects.js : cherche récursivement, à
// partir de rootDir, les répertoires contenant un fichier marqueur de
// projet. Une fois un marqueur trouvé dans un répertoire, ses
// sous-répertoires ne sont pas explorés.
public class FindProjects {

    public static final Map<String, String> PROJECT_MARKERS = Map.of(
        "pom.xml", "maven",
        "package.json", "npm",
        "Cargo.toml", "rust",
        "go.mod", "go"
    );

    private static final Set<String> DEFAULT_IGNORE_DIRS = Set.of(
        "node_modules", "target", "dist", "build", "out", ".git",
        "venv", ".venv", "env", "__pycache__"
    );

    public record FoundProject(Path dir, List<String> files) {}

    public static List<FoundProject> findProjects(Path rootDir) {
        return findProjects(rootDir, Set.of());
    }

    public static List<FoundProject> findProjects(Path rootDir, Set<String> extraIgnoreDirs) {
        Set<String> ignoreDirs = new LinkedHashSet<>(DEFAULT_IGNORE_DIRS);
        ignoreDirs.addAll(extraIgnoreDirs);

        List<FoundProject> results = new ArrayList<>();
        walk(rootDir, ignoreDirs, results);
        return results;
    }

    private static void walk(Path dir, Set<String> ignoreDirs, List<FoundProject> results) {
        List<String> foundMarkers = new ArrayList<>();
        List<Path> subDirs = new ArrayList<>();

        try (DirectoryStream<Path> stream = Files.newDirectoryStream(dir)) {
            for (Path entry : stream) {
                String name = entry.getFileName().toString();
                if (Files.isRegularFile(entry) && PROJECT_MARKERS.containsKey(name)) {
                    foundMarkers.add(name);
                } else if (Files.isDirectory(entry) && !ignoreDirs.contains(name)) {
                    subDirs.add(entry);
                }
            }
        } catch (IOException e) {
            return;
        }

        if (!foundMarkers.isEmpty()) {
            results.add(new FoundProject(dir, foundMarkers));
            return;
        }

        for (Path subDir : subDirs) {
            walk(subDir, ignoreDirs, results);
        }
    }
}
